### PR TITLE
Resolve template schema

### PR DESCRIPTION
**Pull request by mingwli**
1. resolve from template directory by naming convention
   - *.schema.json
2. put default test schemas in test/data/template
3. keep isLandingData static 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4789)

### CODE REVIEW

File: src/docfx/build/legacy/LegacyFileMap.cs [hunk 1]
```diff
@@ -34,7 +34,7 @@ internal static class LegacyFileMap
                     documents,
                     document =>
                     {
-                        if (document.IsSchemaData)
+                        if (document.IsData)
```
Comment (OsmondJiang) [inline src/docfx/build/legacy/LegacyFileMap.cs]: may we pass the context here and remove the ‘IsData‘ from document?

File: src/docfx/build/page/BuildPage.cs [hunk 2]
```diff
@@ -203,7 +203,12 @@ private static (List<Error> errors, bool isPage, (OutputMetadata metadata, JObje
         {
             var obj = token as JObject;
 
-            var (schemaValidator, schemaTransformer) = TemplateEngine.GetJsonSchema(file.Mime);
+            if (context.TemplateEngine is null)
+            {
+                throw Errors.TemplateNotDefined(file).ToException();
+            }
+
+            var (schemaValidator, schemaTransformer) = context.TemplateEngine.GetJsonSchema(file.Mime);
```
Comment (OsmondJiang) [inline src/docfx/build/page/BuildPage.cs]:
```
var (schemaValidator, schemaTransformer) = context.TemplateEngine?.GetJsonSchema(file.Mime);
```

File: src/docfx/docfx.csproj [hunk 3]
```diff
@@ -42,4 +42,8 @@
     <None Include="data/**/*" CopyToOutputDirectory="PreserveNewest" />
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (OsmondJiang) [inline src/docfx/docfx.csproj]: remove?

File: src/docfx/build/legacy/manifest/LegacyManifest.cs [hunk 4]
```diff
@@ -33,7 +33,7 @@ public static void Convert(Docset docset, Context context, Dictionary<Document,
 
                         var output = new LegacyManifestOutput
                         {
-                            MetadataOutput = document.IsSchemaData || document.ContentType == ContentType.Resource
+                            MetadataOutput = document.IsData || document.ContentType == ContentType.Resource
```
Comment (OsmondJiang) [inline src/docfx/build/legacy/manifest/LegacyManifest.cs]: Seems IsData is mostly used in legacy, can we remove it from Document and directly use TemplateEngine.IsData?

File: docs/specs/escape.yml [hunk 5]
```diff
@@ -111,7 +111,8 @@ outputs:
 ---
 # Handle encoding in xref resolving
 inputs:
-  docfx.yml:
+  docfx.yml: |
```
Comment (OsmondJiang) [inline docs/specs/escape.yml]: add it to test global config?

File: src/docfx/build/legacy/manifest/LegacyManifest.cs (hunk 4 above)
Comment (mingwli) [reply src/docfx/build/legacy/manifest/LegacyManifest.cs]: It's also used for v3 to decide sitePath and won't get changed after document is created. I think let document carry it is better.

---
In reply to: [296725063](https://github.com/dotnet/docfx/pull/4789#discussion_r296725063) [](ancestors = 296725063)

File: src/docfx/build/page/BuildPage.cs (hunk 2 above)
Comment (mingwli) [reply src/docfx/build/page/BuildPage.cs]: Removed 'TemplateNotDefined', the templateEngine is never null in context.

---
In reply to: [296721036](https://github.com/dotnet/docfx/pull/4789#discussion_r296721036) [](ancestors = 296721036)

File: src/docfx/build/legacy/manifest/LegacyManifest.cs (hunk 4 above)
Comment (OsmondJiang) [reply src/docfx/build/legacy/manifest/LegacyManifest.cs]: sitePath is calculated during documentation creation, no need document to carry it?

File: src/docfx/lib/markdown/MarkdownUtility.cs [hunk 6]
```diff
@@ -164,7 +164,7 @@ private static MarkdownPipeline CreateTocMarkdownPipeline()
 
         private static string GetToken(string key)
         {
-            return t_status.Value.Peek().Context.Template?.GetToken(key);
+            return t_status.Value.Peek().Context.TemplateEngine?.GetToken(key);
```
Comment (yufeih) [inline src/docfx/lib/markdown/MarkdownUtility.cs]:
```suggestion
            return t_status.Value.Peek().Context.TemplateEngine.GetToken(key);
```

File: src/docfx/template/TemplateEngine.cs [hunk 7]
```diff
@@ -16,40 +16,44 @@ namespace Microsoft.Docs.Build
...
         {
             var contentTemplateDir = Path.Combine(templateDir, "ContentTemplate");
+            var schemaDir = Path.Combine(contentTemplateDir, "schemas");
 
             _templateDir = templateDir;
+            _schemaDir = schemaDir;
             _liquid = new LiquidTemplate(templateDir);
             _js = new JavascriptEngine(contentTemplateDir);
             Global = LoadGlobalTokens(contentTemplateDir);
+            _schemas = Directory.Exists(schemaDir) ? new HashSet<string>(Directory.EnumerateFiles(schemaDir, "*.schema.json", SearchOption.TopDirectoryOnly)
```
Comment (yufeih) [inline src/docfx/template/TemplateEngine.cs]:
```suggestion
            _schemas = Directory.Exists(schemaDir) ? xxx.ToHashSet(PathUtility.PathComparer);
```

File: test/docfx.Test/e2e/E2ETest.cs [hunk 8]
```diff
@@ -23,7 +23,7 @@ public static class E2ETest
     {
         private static readonly string[] s_errorCodesWithoutLineInfo =
         {
-            "need-restore", "heading-not-found", "config-not-found", "committish-not-found", "custom-404-page",
+            "need-restore", "heading-not-found", "config-not-found", "committish-not-found", "custom-404-page", "template-not-defined",
```
Comment (yufeih) [inline test/docfx.Test/e2e/E2ETest.cs]: revert this line

File: docs/specs/rendering.yml [hunk 9]
```diff
@@ -1,5 +1,5 @@
 ---
-# Apply template and produce html page when output is not json
+# [skip] Apply template and produce html page when output is not json
```
Comment (mingwli) [inline docs/specs/rendering.yml]: > [skip] Apply te [](start = 1, length = 16)

skipped for using liquid rendering in v3 build

File: docs/specs/metadata.yml [hunk 10]
```diff
@@ -358,6 +358,7 @@ inputs:
     name: azure-documents
     files: "**/*"
     baseUrl: https://docs.com/articles
+    template: {APP_BASE_PATH}data/template
```
Comment (yufeih) [inline docs/specs/metadata.yml]: template is not needed for tests that only consumes `LandingData`

File: src/docfx/build/dependency/DependencyResolver.cs [hunk 11]
```diff
@@ -245,17 +248,17 @@ internal class DependencyResolver
...
                     }
 
-                    var file = Document.CreateFromFile(declaringFile.Docset, pathToDocset);
+                    var file = Document.CreateFromFile(declaringFile.Docset, pathToDocset, _templateEngine);
 
-                    // forLandingPage should not be used, it is a hack to handle some specific logic for landing page based on the user input for now
+                    // for LandingPage should not be used, it is a hack to handle some specific logic for landing page based on the user input for now
                     // which needs to be removed once the user input is correct
-                    if (TemplateEngine.IsLandingData(declaringFile.Mime))
+                    if (_templateEngine != null && TemplateEngine.IsLandingData(declaringFile.Mime))
```
Comment (yufeih) [inline src/docfx/build/dependency/DependencyResolver.cs]:
```suggestion
                    if ( TemplateEngine.IsLandingData(declaringFile.Mime))
```

File: src/docfx/template/TemplateEngine.cs [hunk 12]
```diff
@@ -16,40 +16,53 @@ namespace Microsoft.Docs.Build
...
-                .Where(prop => !string.IsNullOrEmpty(prop.Value.HtmlMetaName))
-                .ToDictionary(prop => prop.Key, prop => prop.Value.HtmlMetaName);
+            _schemas = Directory.Exists(schemaDir) ? new HashSet<string>(Directory.EnumerateFiles(schemaDir, "*.schema.json", SearchOption.TopDirectoryOnly)
+                                                                        .Select(k => Path.GetFileNameWithoutExtension(k))
+                                                                        .Select(k => k.Substring(0, k.Length - ".schema".Length))) : new HashSet<string>();
+            foreach (var schema in s_schemas)
+            {
+                _schemas.Add(schema);
+            }
+            if (metadataSchema != null)
```
Comment (yufeih) [inline src/docfx/template/TemplateEngine.cs]: nit: prefer an empty `JsonSchema` over `null`

File: src/docfx/template/TemplateEngine.cs [hunk 13]
```diff
@@ -16,40 +16,53 @@ namespace Microsoft.Docs.Build
...
     internal class TemplateEngine
     {
-        // todo: read schema from template
-        private static readonly HashSet<string> s_schemas = new HashSet<string>(Directory.EnumerateFiles(Path.Combine(AppContext.BaseDirectory, "data"), "*.json", SearchOption.TopDirectoryOnly).Select(k => Path.GetFileNameWithoutExtension(k)));
-
         private static readonly string[] s_resourceFolders = new[] { "global", "css", "fonts" };
         private static readonly ConcurrentDictionary<string, Lazy<(JsonSchemaValidator, JsonSchemaTransformer)>> _jsonSchemas
                           = new ConcurrentDictionary<string, Lazy<(JsonSchemaValidator, JsonSchemaTransformer)>>();
 
+        private static readonly string s_systemSchemasDir = Path.Combine(AppContext.BaseDirectory, "data", "schemas");
```
Comment (yufeih) [inline src/docfx/template/TemplateEngine.cs]: nit: can specialize `LandingData` in each function

File: src/docfx/template/TemplateEngine.cs [hunk 14]
```diff
@@ -96,7 +108,7 @@ public static TemplateEngine Create(Docset docset)
 
             if (string.IsNullOrEmpty(docset.Config.Template))
             {
-                return null;
+                return new TemplateEngine("", null);
```
Comment (yufeih) [inline src/docfx/template/TemplateEngine.cs]: Is there a default template path in v2? Something like `{docsetDir}/_theme` @superyyrrzz

**Comment by OsmondJiang** [comment]
check the integration test diff

**Pull request merged**